spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultTaskExecutionService.java

```diff
@@ -151,23 +165,20 @@ public DefaultTaskExecutionService(LauncherRepository launcherRepository,
 		this.taskAppDeploymentRequestCreator = taskAppDeploymentRequestCreator;
 		this.taskExplorer = taskExplorer;
 		this.dataflowTaskExecutionDao = dataflowTaskExecutionDao;
+		this.dataflowTaskExecutionMetadataDao = dataflowTaskExecutionMetadataDao;
 	}
 
 	@Override
 	public long executeTask(String taskName, Map<String, String> taskDeploymentProperties, List<String> commandLineArgs,
 			String composedTaskRunnerName) {
 
-		String platformName = taskDeploymentProperties.get(TASK_PLATFORM_NAME);
+		String platformName = getPlatform(taskDeploymentProperties);
 
-		// If not given, use 'default'
-		if (!StringUtils.hasText(platformName)) {
-			platformName = "default";
-		}
-		// In case we have exactly one launcher, we override to that
-		List<String> launcherNames = StreamSupport.stream(launcherRepository.findAll().spliterator(), false)
-				.map(Launcher::getName).collect(Collectors.toList());
-		if (launcherNames.size() == 1) {
-			platformName = launcherNames.get(0);
+		if(this.tasksBeingUpgraded.containsKey(taskName)) {
```

Comment:
Keeping a local state like this is confusing. It has to lead race conditions at some point and its at least unclear to me how `validateAndLockUpgrade()` works and clear/remove only happens at last steps of this function. What happens when things go south in a middle?

Reply:
You're right on the error condition not being handled.  I'll address that.  The bit about storing this in memory, this is the same mechanism that we use for handling max task executions allowed so I didn't want to make this too complex by sticking it in a db or related store.  I'm open to ideas though.

Reply:
On a skipper side stream state is handled via statemachine and while stream is in either install, upgrade, rollback or delete, you can't do anything else. State is persisted into a database so it should survive a server restart. I assume if you kick upgrade, restart server, everything breaks up as state is lost. We could look using statemachine in a future but at least for now we need to get this logic a bit more clearer.

Reply:
State machine is extreme overkill fo what we're trying to accomplish here IMHO.  Also, keep in mind, the deployment mechanism for tasks is completely different.  There are two specific steps in the upgrade of a task.  delete and launch.  If delete fails...manual intervention will be required anyways.  If the launch fails, you can just run it again.  I don't want to over engineer this if it doesn't need to be that complex.